test: add coverage for resolving TASK_LISTENER_NO_RETRIES error

kristinkomschow opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
Test for operate resolving incidents for task listener jobs.

The test creates an incident by starting a process with a task listener ("completing"), triggering the "complete" action for the user task with the listener and then failing the activated listener job via the zeebe API.

The incident is resolved via the `batchOperationWriter` instead of the Operate API because of authorization issues (we just directly call the method that would be triggered by the REST API after authorization checks) because of upcoming changes to how permissions work (related to [restructuring Identity](https://camunda.slack.com/archives/C06F0GLJNFM/p1736786939932799?thread_ts=1734620941.143069&cid=C06F0GLJNFM) and streamlining permissions in Camunda).

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes https://github.com/camunda/camunda/issues/25451

ThorbenLindhauer reviewed (changes requested):
The test flow overall looks good to me. I have some remarks about the implementation. The most important point for me is the next one here, i.e. the test runtime:

* The test fails locally for me with
```
[ERROR] Errors:
[ERROR]   ResolveIncidentsViaOperateIT.shouldResolveIncident:107 ┬╗ ConditionTim
eout Condition with alias 'should find no more incident for process ' didn't com
plete within 1 minutes because assertion condition defined as a Lambda expressio
n in io.camunda.it.utils.CamundaClientTestRequestHelper
Expecting value to be false but was true.
```
Might just be a timing problem, as [in the PR CI](https://github.com/camunda/camunda/actions/runs/13111526290/job/36576341011) I can also see that it takes more than two minutes to run. In general, should we have a test that runs this long and creates risk of local failure? Can we find a way to test this more efficiently (maybe with reduced scope)?
* Can you please add the ticket reference to the commit messages?

kristinkomschow commented:
I removed the integration test again and essentially replaced it with a unit test because it was slow and we have come to the conclusion that we can cover what we actually want to test here through unit tests and existing integration tests.

I've adjusted the issue description to fit the actual requirement of what we want to test here: https://github.com/camunda/camunda/issues/25451#issue-2704486246

Comment on how we cover the different steps: https://github.com/camunda/camunda/issues/25451#issuecomment-2642916528

ThorbenLindhauer commented on operate/webapp/src/test/java/io/camunda/operate/webapp/zeebe/operation/ResolveIncidentHandlerTest.java:
```diff
@@ -0,0 +1,91 @@
+    when(camundaClient.newUpdateRetriesCommand(jobKey)).thenReturn(updateRetriesCommand);
+    when(camundaClient.newResolveIncidentCommand(incidentKey)).thenReturn(resolveIncidentCommand);
+
+    doNothing().when(resolveIncidentHandler).markAsSent(any());
```
I think this interferes too much with the implementation. Let's not modify the behavior of the class under test. Instead, rather mock the metrics and batch operation writer.

ThorbenLindhauer commented on operate/webapp/src/test/java/io/camunda/operate/webapp/zeebe/operation/ResolveIncidentHandlerTest.java:
```diff
@@ -0,0 +1,91 @@
+    // when
+    resolveIncidentHandler.handleWithException(operationEntity);
+
+    final InOrder inOrder = Mockito.inOrder(camundaClient);
```
Move this into the `then` block. `when` should only contain the statement under test. That makes it easiest to understand what a test really covers.

ThorbenLindhauer commented on qa/integration-tests/src/test/java/io/camunda/it/exporter/ExporterTestUtil.java:
```diff
@@ -134,6 +134,11 @@ public static String startProcessInstance(
 
   public static void waitForProcessTasks(
       final CamundaClient client, final String processInstanceKey) {
+    waitForProcessTasks(client, Long.valueOf(processInstanceKey));
```
Are the changes to this class still relevant?

ThorbenLindhauer commented on qa/util/src/main/java/io/camunda/qa/util/cluster/TestStandaloneCamunda.java:
```diff
@@ -133,9 +133,11 @@ public TestStandaloneCamunda start() {
     esContainer.start();
 
     final String esURL = String.format("http://%s", esContainer.getHttpHostAddress());
+    final String gatewayURL = String.format("localhost:%s", gatewayConfig().getNetwork().getPort());
```
Are the changes to this class still relevant?

ThorbenLindhauer reviewed (changes requested):
Looks good to me overall. Some smaller remarks in the details.

Don't forget to squash your commits so that there are no commits that first add the IT and then later remove it again.

ThorbenLindhauer reviewed (approved):
Looks good, thank you for the quick resolution.
